File: templates/websocket_template/src/core_specs/data/data_loader.py
"""
#############################################################################
### Data loader file
###
### @file data_loader.py
### @author Sebastian Russo
### @date 2025
#############################################################################

This module loads general data from a JSON file.
"""

# Native imports
import json
import sys


def read_data_from_data_json(file_path: str, exit_on_error: bool = True) -> dict:
    """
    Reads data from a JSON data file.

    Parameters:
        file_path (str): Path to the JSON data file.
        exit_on_error (bool): Whether to exit on error or return None.

    Returns:
        dict: Parsed JSON data.
    """
    try:
        with open(file_path, "r", encoding="utf-8") as file:
            data = json.load(file)
        return data
    except FileNotFoundError:
        print(f"ERROR: Data file not found: {file_path}")
        if exit_on_error:
            sys.exit(1)
        return None
    except json.JSONDecodeError:
        print(f"ERROR: Failed to parse JSON data file: {file_path}")
        if exit_on_error:
            sys.exit(1)
        return None


DATA_FILE_PATH = "src/core_specs/data/general_data.json"
data_loader = read_data_from_data_json(DATA_FILE_PATH, exit_on_error=True)

File: templates/java_rest_api_template/src/main/java/com/template/restapi/core_specs/data/DataLoader.java
/**
 * #############################################################################
 * ### Data loader file
 * ###
 * ### @file DataLoader.java
 * ### @author Sebastian Russo
 * ### @date 2026
 * #############################################################################
 *
 * Loads general (non-environment-specific) reference data from
 * general_data.json on the classpath. Direct equivalent of the Python
 * template's data_loader.py.
 *
 * Use it for static lookup tables such as supported languages:
 *   Python:  data_loader["languages"]
 *   Java:    DataLoader.get().get("languages")
 */
package com.template.restapi.core_specs.data;

import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

public final class DataLoader {

    private static final String DATA_PATH = "/core_specs/data/general_data.json";

    private static final JsonNode DATA = load();

    private DataLoader() { }

    private static JsonNode load() {
        try (InputStream in = DataLoader.class.getResourceAsStream(DATA_PATH)) {
            if (in == null) {
                System.err.println("ERROR: Data file not found on classpath: " + DATA_PATH);
                System.exit(1);
            }
            return new ObjectMapper().readTree(in);
        } catch (Exception ex) {
            System.err.println("ERROR: Failed to parse JSON data file '" + DATA_PATH + "': " + ex.getMessage());
            System.exit(1);
            return null; // unreachable
        }
    }

    /** The entire parsed data tree (equivalent to the data_loader dict). */
    public static JsonNode get() {
        return DATA;
    }

    /** Convenience helper: read a top-level string array (e.g. "languages"). */
    public static List<String> stringList(String key) {
        List<String> out = new ArrayList<>();
        JsonNode node = DATA.path(key);
        if (node.isArray()) {
            node.forEach(item -> out.add(item.asText()));
        }
        return out;
    }
}
